### PR TITLE
*: Add ID to IndexInfo and use IndexID in prefix instead of IndexName

### PR DESCRIPTION
**Pull request by shenli**

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -732,7 +736,7 @@ func (d *ddl) buildIndex(ctx context.Context, t table.Table, idxInfo *model.Inde
 			vals = append(vals, val)
 		}
 		// build index
-		kvX := kv.NewKVIndex(t.IndexPrefix(), idxInfo.Name.L, unique)
+		kvX := kv.NewKVIndex(t.IndexPrefix(), idxInfo.Name.L, idxInfo.ID, unique)
```
Comment (ngaut) [inline ddl/ddl.go]: Do we still need name of the index ?

**Comment by siddontang** [comment]
where to generate the index ID?

File: ddl/ddl.go [hunk 2]
```diff
@@ -378,6 +378,10 @@ func (d *ddl) buildTableInfo(tableName model.CIStr, cols []*column.Col, constrai
 		case coldef.ConstrUniq, coldef.ConstrUniqKey, coldef.ConstrUniqIndex:
 			idxInfo.Unique = true
 		}
+		idxInfo.ID, err = d.genGlobalID()
```
Comment (shenli) [inline ddl/ddl.go]: @siddontang Index ID is generated here.

File: ddl/ddl.go (hunk 1 above)
Comment (siddontang) [inline ddl/ddl.go]: for error log later maybe, like ErrKeyExist, we can returns the index name and index value in error for debug easily.

File: kv/index_iter.go [hunk 3]
```diff
@@ -97,26 +97,28 @@ func (c *indexIter) Next() (k []interface{}, h int64, err error) {
...
 
-func genIndexPrefix(indexPrefix, indexName string) string {
+func genIndexPrefix(indexPrefix string, indexID int64) string {
 	// Use EncodeBytes to guarantee generating different index prefix.
 	// e.g, two indices c1 and c with index prefix p, if no EncodeBytes,
 	// the index format looks p_c and p_c1, if c has an index value which the first encoded byte is '1',
 	// we will meet an error, because p_c1 is for index c1.
 	// If EncodeBytes, c1 -> c1\x00\x01 and c -> c\x00\x01, the prefixs are different.
-	key := fmt.Sprintf("%s_%s", indexPrefix, indexName)
+	key := fmt.Sprintf("%s_%d", indexPrefix, indexID)
```
Comment (disksing) [inline kv/index_iter.go]: How about using `codec.EncodeKey()` to make target key more compact?

Comment (qiuyesuifeng) [inline kv/index_iter.go]: +1, if we use index id instead of index name, seems can directly encode?

Comment (siddontang) [inline kv/index_iter.go]: The codec may be refactored later.

Comment (ngaut) [inline kv/index_iter.go]: But we  should use EncodeKey.

Comment (shenli) [inline kv/index_iter.go]: Index prefix use the same way with record prefix.
Maybe in the next pr.

**Comment by shenli** [comment]
Merge master and solve conflicts.
PTAL

File: ddl/ddl.go (hunk 2 above)
Comment (siddontang) [inline ddl/ddl.go]: you should add index ID in CreateIndex DDL too, maybe in buildIndexInfo.

Comment (shenli) [inline ddl/ddl.go]: PTAL

**Comment by shenli** [comment]
PTAL @siddontang

**Comment by siddontang** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**